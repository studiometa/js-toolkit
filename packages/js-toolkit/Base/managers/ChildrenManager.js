import { getComponentElements } from '../utils.js';

/**
 * @typedef {import('../index.js').default} Base
 * @typedef {import('../index.js').BaseConstructor} BaseConstructor
 * @typedef {import('../index.js').BaseAsyncConstructor} BaseAsyncConstructor
 * @typedef {import('../index.js').BaseConfigComponents} BaseConfigComponents
 * @typedef {import('./EventsManager.js').default} EventsManager
 */

/**
 * Children manager.
 */
export default class ChildrenManager {
  /**
   * @type {Base}
   * @private
   */
  __base;

  /**
   * @type {HTMLElement}
   * @private
   */
  __element;

  /**
   * @type {BaseConfigComponents}
   * @private
   */
  __components;

  /**
   * @type {EventsManager}
   * @private
   */
  __eventsManager;

  /**
   * @return {string[]}
   */
  get registeredNames() {
    return Object.keys(this).filter((key) => !key.startsWith('__'));
  }

  /**
   * Class constructor.
   * @param {Base} base
   *   The component's instance.
   * @param {HTMLElement} element
   *   The component's root element
   * @param {BaseConfigComponents} components
   *   The children components' classes
   * @param {EventsManager} eventsManager
   */
  constructor(base, element, components, eventsManager) {
    Object.defineProperties(this, {
      __base: {
        enumerable: false,
        writable: false,
        value: base,
      },
      __element: {
        enumerable: false,
        writable: false,
        value: element,
      },
      __components: {
        enumerable: false,
        writable: false,
        value: components,
      },
      __eventsManager: {
        enumerable: false,
        writable: false,
        value: eventsManager,
      },
    });
  }

  /**
   * Register instances of all children components.
   */
  registerAll() {
    Object.entries(this.__components).forEach(([name, component]) =>
      this.__register(name, component)
    );
  }

  /**
   * Register instance of a child component.
   *
   * @param {string} name
   *   The name of the child component.
   * @param {BaseConstructor|BaseAsyncConstructor} component
   *   A Base class or a Promise for async components.
   * @private
   */
  __register(name, component) {
    Object.defineProperty(this, name, {
      enumerable: true,
      configurable: true,
      get: () => {
        const elements = getComponentElements(name, this.__element);

        if (elements.length === 0) {
          return [];
        }

        return elements
          .map((element) => this.__getChild(element, component, name))
          .filter((instance) => instance !== 'terminated');
      },
    });
  }

  /**
   * Get a child component's instance.
   *
   * @param {HTMLElement & { __base__?: WeakMap<BaseConstructor, Base> }} el
   *   The root element of the child component.
   * @param {BaseConstructor|BaseAsyncConstructor} ComponentClass
   *   A Base class or a Promise for async components.
   * @param {string} name
   *   The name of the child component.
   * @return {Base|Promise|'terminated'}
   *   A Base instance or a Promise resolving to a Base instance.
   * @private
   */
  __getChild(el, ComponentClass, name) {
    const asyncComponentPromise = this.__asyncComponentPromises.get(
      /** @type {BaseAsyncConstructor} */ (ComponentClass)
    );

    // Test if we have a constructor and not a promise or if the promise has been resolved
    if (
      '$isBase' in ComponentClass ||
      (asyncComponentPromise && asyncComponentPromise.status === 'resolved')
    ) {
      let ctor = /** @type {BaseConstructor} */ (ComponentClass);

      // Get resolved constructor from weakmap.
      // Only test for existence as the status was checked before.
      if (asyncComponentPromise) {
        ctor = asyncComponentPromise.ctor;
      }

      // Return existing instance if it exists
      if (el.__base__ && el.__base__.has(ctor)) {
        return el.__base__.get(ctor);
      }

      // Return a new instance if the component class is a child of the Base class
      // eslint-disable-next-line new-cap
      const child = new ctor(el);
      Object.defineProperty(child, '$parent', { get: () => this.__base });
      return child;
    }

    const promise = asyncComponentPromise
      ? asyncComponentPromise.promise
      : ComponentClass(this.__base);

    if (!asyncComponentPromise) {
      this.__asyncComponentPromises.set(ComponentClass, {
        promise,
        status: 'pending',
        ctor: undefined,
      });
    }

    // Resolve async components
    return promise.then((module) => {
      // @ts-ignore
      const ctor = module.default ?? module;

      this.__asyncComponentPromises.set(ComponentClass, {
        promise,
        status: 'resolved',
        ctor,
      });

      return this.__getChild(el, ctor, name);
    });
  }

  /**
   * Store async component promises to avoid calling them multiple times and
   * waiting for them when they are already resolved.
   *
   * @private
   * @type {WeakMap<BaseAsyncConstructor, { promise: ReturnType<BaseAsyncConstructor>, status: 'pending'|'resolved', ctor?: BaseConstructor }>}
   */
  __asyncComponentPromises = new WeakMap();

  /**
   * Mount all child component instances.
   */
  mountAll() {
    this.__triggerHookForAll('$mount');
  }

  /**
   * Update all child component instances.
   */
  updateAll() {
    this.__triggerHookForAll('$update');
  }

  /**
   * Destroy all child component instances.
   */
  destroyAll() {
    this.__triggerHookForAll('$destroy');
  }

  /**
   * Execute the given hook for all children instances.
   *
   * @param {'$mount'|'$update'|'$destroy'} hook The hook to execute.
   * @private
   */
  __triggerHookForAll(hook) {
    this.registeredNames.forEach((name) => {
      this[name].forEach((instance) => {
        if (instance instanceof Promise) {
          instance.then((resolvedInstance) => this.__triggerHook(hook, resolvedInstance, name));
        } else {
          this.__triggerHook(hook, instance, name);
        }
      });
    });
  }

  /**
   * Execute the given hook for the given instance.
   *
   * @param {'$mount'|'$update'|'$destroy'} hook The hook to trigger.
   * @param {Base} instance The target instance.
   * @param {string} name The name of the component.
   * @private
   */
  __triggerHook(hook, instance, name) {
    if (hook === '$update' && !instance.$isMounted) {
      // eslint-disable-next-line no-param-reassign
      hook = '$mount';
    }

    if (hook === '$update' || hook === '$destroy') {
      this.__eventsManager.unbindChild(name, instance);
    }

    if (hook === '$update' || hook === '$mount') {
      this.__eventsManager.bindChild(name, instance);
    }

    instance[hook]();
  }
}
