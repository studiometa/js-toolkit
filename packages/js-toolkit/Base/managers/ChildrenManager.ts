import type {
  Base,
  BaseConstructor,
  BaseAsyncConstructor,
  BaseEl,
} from '../index.js';
import AbstractManager from './AbstractManager.js';
import { getComponentElements } from '../utils.js';

/**
 * Get a child component's instance.
 *
 * @param {ChildrenManager} that
 * @param {HTMLElement & { __base__?: WeakMap<BaseConstructor, Base> }} el
 *   The root element of the child component.
 * @param {BaseConstructor|BaseAsyncConstructor} ComponentClass
 *   A Base class or a Promise for async components.
 * @param {string} name
 *   The name of the child component.
 * @returns {Base|Promise<Base | 'terminated'>|'terminated'}
 *   A Base instance or a Promise resolving to a Base instance.
 * @private
 */
function __getChild(
  // eslint-disable-next-line no-use-before-define
  that: ChildrenManager,
  el: BaseEl,
  ComponentClass: BaseConstructor | BaseAsyncConstructor,
  name: string,
): Base | Promise<Base | 'terminated'> | 'terminated' {
  const asyncComponentPromise = that.__asyncComponentPromises.get(ComponentClass as BaseAsyncConstructor);

  // Test if we have a constructor and not a promise or if the promise has been resolved
  if (
    '$isBase' in ComponentClass ||
    (asyncComponentPromise && asyncComponentPromise.status === 'resolved')
  ) {
    let ctor = ComponentClass as BaseConstructor;

    // Get resolved constructor from weakmap.
    // Only test for existence as the status was checked before.
    if (asyncComponentPromise) {
      ctor = asyncComponentPromise.ctor;
    }

    // Return existing instance if it exists
    if (el.__base__ && el.__base__.has(ctor)) {
      return el.__base__.get(ctor);
    }

    // Return a new instance if the component class is a child of the Base class
    // eslint-disable-next-line new-cap
    const child = new ctor(el);
    Object.defineProperty(child, '$parent', { get: () => that.__base });
    return child;
  }

  const promise = asyncComponentPromise
    ? asyncComponentPromise.promise
    : ComponentClass(that.__base);

  if (!asyncComponentPromise) {
    that.__asyncComponentPromises.set(ComponentClass, {
      promise,
      status: 'pending',
      ctor: undefined,
    });
  }

  // Resolve async components
  return promise.then((module) => {
    // @ts-ignore
    const ctor = module.default ?? module;

    that.__asyncComponentPromises.set(ComponentClass, {
      promise,
      status: 'resolved',
      ctor,
    });

    return __getChild(that, el, ctor, name);
  });
}

/**
 * Register instance of a child component.
 *
 * @param {ChildrenManager} that
 * @param {string} name
 *   The name of the child component.
 * @param {BaseConstructor|BaseAsyncConstructor} component
 *   A Base class or a Promise for async components.
 * @private
 */
function __register(
  // eslint-disable-next-line no-use-before-define
  that: ChildrenManager,
  name: string,
  component: BaseConstructor | BaseAsyncConstructor,
) {
  Object.defineProperty(that, name, {
    enumerable: true,
    configurable: true,
    get: () => {
      const elements = getComponentElements(name, that.__element);

      if (elements.length === 0) {
        return [];
      }

      return elements
        .map((element) => __getChild(that, element, component, name))
        .filter((instance) => instance !== 'terminated');
    },
  });
}

/**
 * Execute the given hook for the given instance.
 *
 * @param {ChildrenManager} that
 * @param {'$mount'|'$update'|'$destroy'} hook The hook to trigger.
 * @param {Base} instance The target instance.
 * @param {string} name The name of the component.
 * @private
 */
function __triggerHook(
  // eslint-disable-next-line no-use-before-define
  that: ChildrenManager,
  hook: '$mount' | '$update' | '$destroy',
  instance: Base,
  name: string,
) {
  if (hook === '$update' && !instance.$isMounted) {
    // eslint-disable-next-line no-param-reassign
    hook = '$mount';
  }

  if (hook === '$update' || hook === '$destroy') {
    that.__eventsManager.unbindChild(name, instance);
  }

  if (hook === '$update' || hook === '$mount') {
    that.__eventsManager.bindChild(name, instance);
  }

  instance[hook]();
}

/**
 * Execute the given hook for all children instances.
 *
 * @param {ChildrenManager} that
 * @param {'$mount'|'$update'|'$destroy'} hook The hook to execute.
 * @private
 */
// eslint-disable-next-line no-use-before-define
function __triggerHookForAll(that: ChildrenManager, hook: '$mount' | '$update' | '$destroy') {
  that.registeredNames.forEach((name) => {
    that[name].forEach((instance) => {
      if (instance instanceof Promise) {
        instance.then((resolvedInstance) => __triggerHook(that, hook, resolvedInstance, name));
      } else {
        __triggerHook(that, hook, instance, name);
      }
    });
  });
}

/**
 * Children manager.
 */
export default class ChildrenManager extends AbstractManager {
  /**
   * Store async component promises to avoid calling them multiple times and
   * waiting for them when they are already resolved.
   */
  __asyncComponentPromises: WeakMap<
    BaseAsyncConstructor,
    {
      promise: ReturnType<BaseAsyncConstructor>;
      status: 'pending' | 'resolved';
      ctor?: BaseConstructor;
    }
  > = new WeakMap();

  get registeredNames():string[] {
    return Object.keys(this).filter((key) => !key.startsWith('__'));
  }

  /**
   * Register instances of all children components.
   */
  registerAll() {
    Object.entries(this.__config.components).forEach(([name, component]) =>
      __register(this, name, component),
    );
  }

  /**
   * Mount all child component instances.
   */
  mountAll() {
    __triggerHookForAll(this, '$mount');
  }

  /**
   * Update all child component instances.
   */
  updateAll() {
    __triggerHookForAll(this, '$update');
  }

  /**
   * Destroy all child component instances.
   */
  destroyAll() {
    __triggerHookForAll(this, '$destroy');
  }
}
