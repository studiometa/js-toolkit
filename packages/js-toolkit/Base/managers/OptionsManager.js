import deepmerge from 'deepmerge';
import isObject from '../../utils/object/isObject.js';

/**
 * @typedef {import('deepmerge').Options} DeepmergeOptions
 * @typedef {import('../index.js').BaseConfig} BaseConfig
 * @typedef {StringConstructor|NumberConstructor|BooleanConstructor|ArrayConstructor|ObjectConstructor} OptionType
 * @typedef {{ type: OptionType, default: String|Number|Boolean|(() => Array|Object), merge?: boolean|DeepmergeOptions }} OptionObject
 * @typedef {OptionType | OptionObject} OptionValue
 * @typedef {{ [name:string]: OptionValue }} OptionsSchema
 * @typedef {{ [optionName:string]: any }} OptionsInterface
 */

/**
 * Class options to manage options as data attributes on an HTML element.
 *
 * @todo Use `MutationObserver` to update values? Might be more performant.
 * @augments {OptionsInterface}
 */
export default class OptionsManager {
  /**
   * The HTML element holding the options attributes.
   * @type {HTMLElement}
   * @private
   */
  __element;

  /**
   * An object to store Array and Object values for reference.
   * @type {Object}
   * @private
   */
  __values = {};

  /**
   * List of allowed types.
   * @type {Array}
   * @private
   */
  static types = [String, Number, Boolean, Array, Object];

  /**
   * The default values to return for each available type.
   * @type {Object}
   * @private
   */
  __defaultValues = {
    String: '',
    Number: 0,
    Boolean: false,
    Array: () => [],
    Object: () => ({}),
  };

  /**
   * Default value for the name property.
   * @type {string}
   */
  name = 'Base';

  /**
   * Default value for the debug property.
   * @type {boolean}
   */
  debug = false;

  /**
   * Default value for the log property.
   * @type {Boolean}
   */
  log = false;

  /**
   * Class constructor.
   *
   * @param {HTMLElement}   element The HTML element storing the options.
   * @param {OptionsSchema} schema  A Base class config options.
   * @param {BaseConfig}    baseConfig  A Base class config.
   */
  constructor(element, schema, baseConfig) {
    Object.defineProperties(this, {
      __element: {
        enumerable: false,
        writable: false,
        value: element,
      },
      __values: {
        enumerable: false,
        writable: false,
        value: this.__values,
      },
      __defaultValues: {
        enumerable: false,
        writable: false,
        value: this.__defaultValues,
      },
    });

    this.name = baseConfig.name;

    schema.debug = {
      type: Boolean,
      default: baseConfig.debug ?? false,
    };

    schema.log = {
      type: Boolean,
      default: baseConfig.log ?? false,
    };

    Object.entries(schema).forEach(([name, config]) => {
      this.__register(name, config);
    });

    return this;
  }

  /**
   * Register an option.
   *
   * @param  {string} name
   * @param  {OptionValue} config
   * @return {void}
   * @private
   */
  __register(name, config) {
    const isObjectConfig = typeof config === 'object';
    const hasDefaultValue = !OptionsManager.types.includes(config);
    const type = isObjectConfig
      ? /** @type {OptionObject} */ (config).type
      : /** @type {OptionType} */ (config);

    if (!OptionsManager.types.includes(type)) {
      throw new Error(
        `The "${name}" option has an invalid type. The allowed types are: String, Number, Boolean, Array and Object.`
      );
    }

    // @ts-ignore
    const defaultValue = hasDefaultValue ? config.default : this.__defaultValues[type.name];

    if ((type === Array || type === Object) && typeof defaultValue !== 'function') {
      throw new Error(
        `The default value for options of type "${type.name}" must be returned by a function.`
      );
    }

    Object.defineProperty(this, name, {
      get: () => {
        return this.get(name, type, defaultValue, isObjectConfig ? config.merge : undefined);
      },
      set: (value) => {
        this.set(name, type, value, defaultValue);
      },
      enumerable: true,
    });
  }

  /**
   * Get an option value.
   *
   * @param {String} name The option name.
   * @param {OptionType} type The option data's type.
   * @param {any} defaultValue The default value for this option.
   * @param {boolean|DeepmergeOptions} [merge] Wether to merge the value with the default or not.
   */
  get(name, type, defaultValue, merge) {
    const propertyName = OptionsManager.__getPropertyName(name);
    const hasProperty = this.__hasProperty(propertyName);

    if (type === Boolean) {
      if (defaultValue) {
        const negatedPropertyName = OptionsManager.__getPropertyName(name, 'No');
        const hasNegatedProperty = this.__hasProperty(negatedPropertyName);

        return !hasNegatedProperty;
      }

      return hasProperty || defaultValue;
    }

    const value = this.__element.dataset[propertyName];

    if (type === Number) {
      return hasProperty ? Number(value) : defaultValue;
    }

    if (type === Array || type === Object) {
      if (!this.__values[name]) {
        let val = hasProperty ? JSON.parse(value) : this.__defaultValues[type.name]();

        if (typeof merge !== 'undefined') {
          val =
            typeof merge === 'boolean'
              ? deepmerge(defaultValue(), val)
              : deepmerge(defaultValue(), val, merge);
        }

        this.__values[name] = val;
      }

      return this.__values[name];
    }

    return hasProperty ? value : defaultValue;
  }

  /**
   * Set an option value.
   *
   * @param {String} name The option name.
   * @param {OptionType} type The option data's type.
   * @param {any} value The new value for this option.
   */
  set(name, type, value, defaultValue) {
    const propertyName = OptionsManager.__getPropertyName(name);

    if (value.constructor.name !== type.name) {
      const val = Array.isArray(value) || isObject(value) ? JSON.stringify(value) : value;
      throw new TypeError(
        `The "${val}" value for the "${name}" option must be of type "${type.name}"`
      );
    }

    switch (type) {
      case Boolean:
        if (defaultValue) {
          const negatedPropertyName = OptionsManager.__getPropertyName(name, 'No');

          if (value) {
            delete this.__element.dataset[negatedPropertyName];
          } else {
            this.__element.dataset[negatedPropertyName] = '';
          }
        } else if (value) {
          this.__element.dataset[propertyName] = '';
        } else {
          delete this.__element.dataset[propertyName];
        }
        break;
      case Array:
      case Object:
        this.__values[name] = value;
        break;
      default:
        this.__element.dataset[propertyName] = value;
    }
  }

  /**
   * Test if the element has a given property.
   *
   * @param  {string} prop
   * @return {boolean}
   */
  __hasProperty(prop) {
    return typeof this.__element.dataset[prop] !== 'undefined';
  }

  /**
   * Property name cache.
   * @type {Map}
   * @private
   */
  static __propertyNameCache = new Map();

  /**
   * Get a property name.
   *
   * @param  {string} name
   * @return {string}
   * @private
   */
  static __getPropertyName(name, prefix = '') {
    const key = name + prefix;

    if (OptionsManager.__propertyNameCache.has(key)) {
      return OptionsManager.__propertyNameCache.get(key);
    }

    const propertyName = `option${prefix}${name.replace(/^\w/, (c) => c.toUpperCase())}`;
    OptionsManager.__propertyNameCache.set(key, propertyName);
    return propertyName;
  }
}
