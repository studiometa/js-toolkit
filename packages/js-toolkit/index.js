export { default as Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';
