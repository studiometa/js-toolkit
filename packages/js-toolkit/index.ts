export { Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';

export type {
  BaseTypeParameter,
  BaseConstructor,
  BaseAsyncConstructor,
  BaseConfig,
} from './Base/index.js';

export type { BaseInterface, BaseDecorator } from './Base/types.js';
