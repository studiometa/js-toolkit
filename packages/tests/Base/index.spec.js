/* eslint-disable no-new, require-jsdoc, max-classes-per-file */
import { jest } from '@jest/globals';
import { Base } from '@studiometa/js-toolkit';
import { html } from 'htl';
import ChildrenManager from '@studiometa/js-toolkit/Base/managers/ChildrenManager';
import OptionsManager from '@studiometa/js-toolkit/Base/managers/OptionsManager';
import RefsManager from '@studiometa/js-toolkit/Base/managers/RefsManager';
import wait from '../__utils__/wait';

describe('The abstract Base class', () => {
  it('must be extended', () => {
    expect(() => {
      new Base(document.createElement('div'));
    }).toThrow();
  });

  it('should throw an error when extended without proper configuration', () => {
    expect(() => {
      // @ts-ignore
      class Foo extends Base {
        static config = {};
      }
      new Foo(document.createElement('div'));
    }).toThrow('The `config.name` property is required.');
  });

  it('should throw an error if instantiated without a root element.', () => {
    expect(() => {
      class Foo extends Base {
        static config = {
          name: 'Foo',
        };
      }
      // @ts-ignore
      new Foo();
    }).toThrow('The root element must be defined.');
  });
});

describe('A Base instance', () => {
  class Foo extends Base {
    static config = {
      name: 'Foo',
    };
  }
  const element = document.createElement('div');
  const foo = new Foo(element).$mount();

  it('should have an `$id` property', () => {
    expect(foo.$id).toBeDefined();
  });

  it('should have an `$isMounted` property', () => {
    expect(foo.$isMounted).toBe(true);
  });

  it('should have a `$refs` property', () => {
    expect(foo.$refs).toBeInstanceOf(RefsManager);
  });

  it('should have a `$children` property', () => {
    expect(foo.$children).toBeInstanceOf(ChildrenManager);
  });

  it('should have an `$options` property', () => {
    expect(foo.$options).toBeInstanceOf(OptionsManager);
    expect(foo.$options.name).toBe('Foo');
  });

  it('should be able to set any `$options` property', () => {
    foo.$options.log = true;
    expect(foo.$options.log).toBe(true);
  });

  it('should have an `$el` property', () => {
    expect(foo.$el).toBe(element);
  });

  it('should have a `__base__` property', () => {
    // @ts-ignore
    expect(foo.$el.__base__).toBeInstanceOf(WeakMap);
    expect(foo.$el.__base__.get(Foo)).toBe(foo);
  });

  it('should inherit from parent config', () => {
    class A extends Base {
      static config = {
        name: 'A',
        log: true,
      };
    }

    class B extends A {
      static config = {
        name: 'B',
        options: {
          title: String,
          color: String,
        },
      };
    }

    class C extends B {
      static config = {
        name: 'C',
        options: {
          color: Boolean,
        },
      };
    }

    class D extends C {
      static config = {
        name: 'D',
        log: false,
      };
    }

    const d = new D(document.createElement('div'));
    expect(d.__config).toMatchSnapshot();
  });

  it('should have a `$root` property', () => {
    class ChildComponent extends Base {
      static config = {
        name: 'ChildComponent',
      };
    }

    class Component extends Base {
      static config = {
        name: 'Component',
        components: { ChildComponent },
      };
    }

    class App extends Base {
      static config = {
        name: 'App',
        components: { Component },
      };
    }

    const tpl = html`<div>
      <div data-component="Component">
        <div data-component="ChildComponent"></div>
      </div>
    </div>`;
    const app = new App(tpl).$mount();

    expect(app.$root).toBe(app);
    expect(app.$children.Component[0].$root).toBe(app);
    expect(app.$children.Component[0].$children.ChildComponent[0].$root).toBe(app);
  });
});

describe('A Base instance methods', () => {
  class Foo extends Base {
    static config = {
      name: 'Foo',
    };
  }

  let element;
  let foo;

  beforeEach(() => {
    element = document.createElement('div');
    foo = new Foo(element).$mount();
  });

  it('should emit a mounted event', () => {
    const fn = jest.fn();
    foo.$on('mounted', fn);
    foo.$destroy();
    foo.$mount();
    expect(fn).toHaveBeenCalledTimes(1);
    foo.$mount();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should emit a destroyed event', () => {
    const fn = jest.fn();
    foo.$on('destroyed', fn);
    foo.$destroy();
    expect(fn).toHaveBeenCalledTimes(1);
    foo.$destroy();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should be able to update its child components.', async () => {
    const div = document.createElement('div');
    div.innerHTML = `
      <div data-component="Bar"></div>
      <div data-component="Bar"></div>
    `;

    const fn = jest.fn();
    class Bar extends Foo {}
    class Baz extends Base {
      static config = { name: 'Baz' };
      updated() {
        fn(this.$id);
      }
    }

    class AsyncBaz extends Baz {
      static config = {
        name: 'AsyncBaz',
      };
    }

    class App extends Base {
      static config = {
        name: 'App',
        components: {
          Bar,
          Baz,
          AsyncBaz: () => Promise.resolve(AsyncBaz),
        },
      };
    }

    const app = new App(div).$mount();
    expect(app.$children.Bar).toHaveLength(2);
    expect(app.$children.Bar[0].$isMounted).toBe(true);
    div.innerHTML = `
      <div data-component="Baz"></div>
      <div data-component="Baz"></div>
      <div data-component="AsyncBaz"></div>
    `;

    app.$update();

    expect(app.$children.Bar).toEqual([]);
    expect(app.$children.Baz).toHaveLength(2);
    expect(app.$children.Baz[0].$isMounted).toBe(true);
    const asyncBaz = await app.$children.AsyncBaz[0];
    expect(asyncBaz.$isMounted).toBe(true);

    const id = div.firstElementChild.__base__.get(Baz).$id;
    expect(id).toBe(app.$children.Baz[0].$id);

    const child = document.createElement('div');
    child.setAttribute('data-component', 'Baz');
    div.appendChild(child);
    expect(id).toBe(app.$children.Baz[0].$id);

    app.$update();

    // Wait for the async component to update
    await wait(1);
    expect(id).toBe(app.$children.Baz[0].$id);
    expect(fn).toHaveBeenCalledTimes(3);
  });

  it('should implement the $factory method', () => {
    class Bar extends Foo {}
    class Baz extends Foo {}
    class Boz extends Foo {}
    document.body.innerHTML = `
      <div data-component="Bar"></div>
      <div data-component="Bar"></div>
      <div class="custom-selector"></div>
      <div class="custom-selector"></div>
    `;

    const barInstances = Bar.$factory('Bar');
    const bazInstances = Baz.$factory('.custom-selector');
    const bozInstances = Boz.$factory('Boz');
    expect(barInstances).toHaveLength(2);
    expect(barInstances[0] instanceof Bar).toBe(true);
    expect(bazInstances).toHaveLength(2);
    expect(bazInstances[0] instanceof Baz).toBe(true);
    expect(bozInstances).toHaveLength(0);
    expect(Baz.$factory).toThrow(/\$factory method/);
  });

  it('should be able to be terminated', () => {
    const fn = jest.fn();
    class Bar extends Foo {
      terminated() {
        fn('method');
      }
    }

    const div = document.createElement('div');
    const bar = new Bar(div).$mount();
    expect(bar).toEqual(div.__base__.get(Bar));
    bar.$on('terminated', () => fn('event'));
    bar.$terminate();
    expect(fn).toHaveBeenCalledTimes(2);
    expect(fn).toHaveBeenNthCalledWith(1, 'event');
    expect(fn).toHaveBeenNthCalledWith(2, 'method');
  });

  it('should not find children if none provided', () => {
    class Bar extends Base {
      static config = {
        name: 'Bar',
      };
    }

    class Baz extends Base {
      static config = {
        name: 'Baz',
        components: { Bar },
      };
    }
    expect(foo.$children.registeredNames).toEqual([]);
    expect(new Baz(document.createElement('div')).$mount().$children.Bar).toEqual([]);
  });

  it('should not find terminated children', () => {
    class Bar extends Base {
      static config = {
        name: 'Bar',
      };
    }

    class Baz extends Base {
      static config = {
        name: 'Baz',
        components: { Bar },
      };
    }

    const div = document.createElement('div');
    div.innerHTML = `
      <div data-component="Bar"></div>
    `;
    const baz = new Baz(div).$mount();
    expect(baz.$children.Bar).toEqual([div.firstElementChild.__base__.get(Bar)]);
    div.firstElementChild.__base__.get(Bar).$terminate();
    expect(baz.$children.Bar).toEqual([]);
  });

  it('should listen to the window.onload event', () => {
    const fn = jest.fn();
    class Bar extends Foo {
      loaded() {
        fn();
      }
    }

    const bar = new Bar(document.createElement('div')).$mount();
    window.dispatchEvent(new CustomEvent('load'));
    expect(fn).toHaveBeenCalledTimes(1);
    bar.$destroy();
    window.dispatchEvent(new CustomEvent('load'));
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should mount and destroy its children', () => {
    class Bar extends Base {
      static config = {
        name: 'Bar',
      };
    }
    class Baz extends Foo {
      static config = {
        name: 'Baz',
        components: { Bar },
      };
    }

    document.body.innerHTML = `<div data-component="Bar"></div>`;
    const baz = new Baz(document.body).$mount();
    const barElement = document.querySelector('[data-component="Bar"]');
    expect(baz.$isMounted).toBe(true);
    expect(barElement.__base__.get(Bar).$isMounted).toBe(true);
    baz.$destroy();
    expect(baz.$isMounted).toBe(false);
    expect(barElement.__base__.get(Bar).$isMounted).toBe(false);
  });
});

describe('The Base class event methods', () => {
  it('should bind handlers to events', () => {
    class App extends Base {
      static config = {
        name: 'A',
        emits: ['foo'],
      };
    }

    const app = new App(document.createElement('div')).$mount();
    const fn = jest.fn();
    const off = app.$on('foo', fn);
    app.$emit('foo', { foo: true });
    expect(fn).toHaveBeenCalledTimes(1);
    expect(fn).toHaveBeenLastCalledWith(
      expect.objectContaining({ type: 'foo', detail: [{ foo: true }] })
    );
    off();
    app.$emit('foo', { foo: true });
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should store event handlers', () => {
    class App extends Base {
      static config = { name: 'App', emits: ['event'] };
    }

    const app = new App(document.createElement('div')).$mount();
    const fn = jest.fn();
    app.$on('event', fn);
    expect(app.__hasEvent('event')).toBe(true);
    expect(app.__eventHandlers.get('event')).toEqual(new Set([fn]));
    app.$off('event', fn);
    expect(app.__eventHandlers.get('event')).toEqual(new Set());
  });

  it('should warn when an event is not configured', () => {
    class App extends Base {
      static config = { name: 'App' };
    }

    const app = new App(document.createElement('div')).$mount();
    const warnMock = jest.spyOn(console, 'warn');
    warnMock.mockImplementation(() => undefined);
    app.$on('other-event');
    expect(warnMock).toHaveBeenCalledTimes(1);
    warnMock.mockRestore();
  });
});

describe('A Base instance config', () => {
  let element;

  beforeEach(() => {
    element = document.createElement('div');
  });

  it('should have a working $log method when active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: true,
      };
    }
    const spy = jest.spyOn(window.console, 'log');
    spy.mockImplementation(() => true);
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(true);
    foo.$log('bar');
    expect(spy).toHaveBeenCalledWith('[Foo]', 'bar');
    spy.mockRestore();
  });

  it('should have a silent $log method when not active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: false,
      };
    }
    const spy = jest.spyOn(window.console, 'log');
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(false);
    foo.$log('bar');
    expect(spy).toHaveBeenCalledTimes(0);
    spy.mockRestore();
  });

  it('should have a working debug method when active in dev mode', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        debug: true,
      };
    }
    globalThis.__DEV__ = true;
    process.env.NODE_ENV = 'development';
    const spy = jest.spyOn(window.console, 'log');
    spy.mockImplementation(() => true);
    const div = document.createElement('div');
    const foo = new Foo(div).$mount();
    expect(spy.mock.calls).toMatchSnapshot();
    spy.mockRestore();
    process.env.NODE_ENV = 'test';
  });
});
