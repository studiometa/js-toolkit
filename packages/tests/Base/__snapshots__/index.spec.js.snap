// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Base instance config should have a working debug method when active in dev mode 1`] = `
Array [
  Array [
    "[debug] [Foo-37]",
    "constructor",
    Foo {
      "$el": <div />,
      "$id": "Foo-37",
      "$isMounted": true,
      "$parent": null,
      "__children": ChildrenManager {
        "__asyncComponentPromises": WeakMap {},
      },
      "__eventHandlers": Map {},
      "__events": EventsManager {},
      "__options": OptionsManager {
        "debug": true,
        "log": false,
        "name": "Foo",
      },
      "__refs": RefsManager {},
      "__services": ServicesManager {
        "__customServices": Object {},
      },
    },
  ],
  Array [
    "[debug] [Foo-37]",
    "$emit",
    "before-mounted",
    Array [],
  ],
  Array [
    "[debug] [Foo-37]",
    "$mount",
  ],
  Array [
    "[debug] [Foo-37]",
    "callMethod",
    "mounted",
  ],
  Array [
    "[debug] [Foo-37]",
    "$emit",
    "mounted",
    Array [],
  ],
]
`;

exports[`A Base instance should inherit from parent config 1`] = `
Object {
  "components": Object {},
  "emits": Array [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": Object {
    "color": [Function],
    "title": [Function],
  },
  "refs": Array [],
}
`;
